### PR TITLE
proposal update: add gvk field and clarify about scale validation

### PR DESCRIPTION
**Pull request by nikhita**

### CODE REVIEW

File: contributors/design-proposals/customresources-subresources.md [hunk 1]
```diff
@@ -82,6 +82,9 @@ type CustomResourceSubResourceScale struct {
     // optional, e.g. “.spec.labelSelector”. Must be under `.spec`.
     // Only JSON paths without the array notation are allowed.
     LabelSelectorPath string `json:"labelSelectorPath,omitempty"`
+    // GVK denotes the gvk of the Scale object sent as the payload
+    // for /scale. It allows transition to future versions easily.
+    GVK schema.GroupVersionKind `json:"gvk,omitempty"`
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: Would call it GroupVersionKind, the long version

Comment (sttts) [reply contributors/design-proposals/customresources-subresources.md]: and: Today only `autoscaling/v1.Scale` (is this the official syntax?) is allowed. If left empty, it default to this value.

File: contributors/design-proposals/customresources-subresources.md [hunk 2]
```diff
@@ -94,6 +97,10 @@ If the gate is not open, the value of the new field within `CustomResourceDefini
 
 The `Scale` object is the payload sent over the wire for `/scale`. The [polymorphic `Scale` type](https://github.com/kubernetes/kubernetes/pull/53743) i.e. `autoscaling/v1.Scale` is used for the `Scale` object.
 
+Since the GVK of the `Scale` object is specified in `CustomResourceSubResourceScale`, transition to future versions (eg `autoscaling/v2.Scale`) can be done easily.
+
+Note: If `v1` is deprecated in `autoscaling`, then it would be deprecated here as well.
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: `autoscaling/v1.Scale`

File: contributors/design-proposals/customresources-subresources.md [hunk 3]
```diff
@@ -123,7 +130,11 @@ Moreover, if the scale subresource is enabled:
 - The value at the optional JSON Path `StatusReplicasPath` (e.g. `.status.replicas`) is validated to be a an integer number if it exists (i.e. this can be empty).
 
 - The value at the optional JSON Path `LabelSelectorPath` (e.g. `.spec.labelSelector`) is validated to be a valid label selector if it exists (i.e. this can be empty).
-If it is invalid, an empty `LabelSelector` is used.
+
+If the `CustomResourceSubResourceScale` field:
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: If one of the path fields in `CustomResourceSubResourceScale`...

File: contributors/design-proposals/customresources-subresources.md [hunk 4]
```diff
@@ -123,7 +130,11 @@ Moreover, if the scale subresource is enabled:
...
+- does not exist, it maps to the default value (0 for the replicas and empty for the label selector).
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: does not exist within the CR..., same in the next line

File: contributors/design-proposals/customresources-subresources.md [hunk 5]
```diff
@@ -127,16 +127,16 @@ validation:
 
 Moreover, if the scale subresource is enabled:
 
-- The value at the specified JSON Path `SpecReplicasPath` (e.g. `.spec.replicas`) is validated to ensure that it contains non-negative integer value and is not empty.
+On update, we copy the values from the `Scale` object into the specified paths in the CustomResource, if the path is set (`StatusReplicasPath` and `LabelSelectorPath` are optional).
+If `StatusReplicasPath` or `LabelSelectorPath` is not set, we validate that the value in `Scale` is also not specified and return an error otherwise.
 
-- The value at the optional JSON Path `StatusReplicasPath` (e.g. `.status.replicas`) is validated to be a an integer number if it exists (i.e. this can be empty).
+On get and on update (after copying the value into the CustomResource as described above), we verify that:
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: On `get` and on `update`

Comment (sttts) [reply contributors/design-proposals/customresources-subresources.md]: values

Comment (nikhita) [reply contributors/design-proposals/customresources-subresources.md]: done

File: contributors/design-proposals/customresources-subresources.md [hunk 6]
```diff
@@ -127,16 +127,16 @@ validation:
...
 
-- The value at the optional JSON Path `LabelSelectorPath` (e.g. `.spec.labelSelector`) is validated to be a valid label selector if it exists (i.e. this can be empty).
+- The value at the specified JSON Path `SpecReplicasPath` (e.g. `.spec.replicas`) is a non-negative integer value and is not empty.
 
-If one of the path fields in `CustomResourceSubResourceScale`:
-- does not exist within the CR, it maps to the default value (0 for the replicas and empty for the label selector).
-- exists within the CR but it does not validate (no integer or no label selector), we return an error.
+- The value at the optional JSON Path `StatusReplicasPath` (e.g. `.status.replicas`) is an integer value if it exists (i.e. this can be empty).
 
+- The value at the optional JSON Path `LabelSelectorPath` (e.g. `.spec.labelSelector`) is a valid label selector if it exists (i.e. this can be empty).
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: I wonder, the same validation should be done when the whole CR is updated and when .status is updated. Maybe we just add this here with another sentence.

Comment (nikhita) [reply contributors/design-proposals/customresources-subresources.md]: won't we end up validating spec with status updates then?

Comment (nikhita) [reply contributors/design-proposals/customresources-subresources.md]: but we should mention at least about `statusReplicasPath`

File: contributors/design-proposals/customresources-subresources.md [hunk 7]
```diff
@@ -138,6 +138,8 @@ On `get` and on `update` (after copying the values into the CustomResource as de
 
 - The value at the optional JSON Path `LabelSelectorPath` (e.g. `.spec.labelSelector`) is a valid label selector if it exists (i.e. this can be empty).
 
+**Note**: The values at the JSON Paths specified by `SpecReplicasPath`, `LabelSelectorPath` and `StatusReplicasPath` also validated with the same rules when the whole object or, in case of an enable of the `/status` subresource, the `.status` sub-object is updated.
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: are also validated

Comment (sttts) [reply contributors/design-proposals/customresources-subresources.md]: enabled

File: contributors/design-proposals/customresources-subresources.md [hunk 8]
```diff
@@ -82,6 +82,11 @@ type CustomResourceSubResourceScale struct {
     // optional, e.g. “.spec.labelSelector”. Must be under `.spec`.
     // Only JSON paths without the array notation are allowed.
     LabelSelectorPath string `json:"labelSelectorPath,omitempty"`
+    // GVK denotes the gvk of the Scale object sent as the payload
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: GroupVersionKind

File: contributors/design-proposals/customresources-subresources.md [hunk 9]
```diff
@@ -94,6 +99,10 @@ If the gate is not open, the value of the new field within `CustomResourceDefini
 
 The `Scale` object is the payload sent over the wire for `/scale`. The [polymorphic `Scale` type](https://github.com/kubernetes/kubernetes/pull/53743) i.e. `autoscaling/v1.Scale` is used for the `Scale` object.
 
+Since the GVK of the `Scale` object is specified in `CustomResourceSubResourceScale`, transition to future versions (eg `autoscaling/v2.Scale`) can be done easily.
+
+Note: If `v1` is deprecated in `autoscaling/v1.Scale`, then it would be deprecated here as well.
```
Comment (sttts) [inline contributors/design-proposals/customresources-subresources.md]: If `autoscaling/v1.Scale` is deprecated, ...

Comment (nikhita) [reply contributors/design-proposals/customresources-subresources.md]: done

File: contributors/design-proposals/customresources-subresources.md (hunk 8 above)
Comment (nikhita) [reply contributors/design-proposals/customresources-subresources.md]: done

**Pull request closed without merging**